### DISCUSSION TITLE
v3.6.0

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#21717] [Announcements]
> 3.6.0 is the next minor release, packed with improvements and bug fixes.

## 📣 Upcoming news

In the coming week you can expect two announcements:

1. The open-sourcing of `nuxt/cli` by [**@pi0**](https://github.com/pi0) - a new, drop-in replacement for `nuxi` featuring more extensibility and better DX. We are aiming to release this alongside Nuxt 3.7, but you would be very welcome to test and contribute to `nuxi-ng` before then!
2. Three significant RFCs: **Nuxt Font**, **Nuxt Assets** and **Nuxt Scripts**. We're expecting an initial POC to take shape in the next month, and would very much appreciate community feedback on this.

## 👀 Highlights

This minor release contains quite a lot, and we have big plans 

### <img src="https://nuxt.com/assets/design-kit/logo/icon-green.svg" width="16" height="16" valign="center"> &nbsp;SPA loading indicator

If your site is served with `ssr: false` or you have disabled server-rendering on some of your pages, you might be particularly interested in the new [built-in SPA loading indicator](https://github.com/nuxt/nuxt/pull/21640).

You can now place an HTML file in `~/app/spa-loading-template.html` with some HTML you would like to use to render a loading screen that will be rendered until your app is hydrated on these pages.

👉 **By default an animated Nuxt icon is rendered**. You can completely disable this indicator by setting `spaLoadingTemplate: false` in your nuxt configuration file.

### ⚡️ Performance improvements

The first thing that happens when your app is hydrated is that your plugins run, and so we now perform [build-time optimisations on your plugins](https://github.com/nuxt/nuxt/pull/21611), meaning they do not need to be normalised or reordered at runtime.

We also include your error component JS in your main entrypoint, meaning that if an error occurs when a user has no connectivity, you can still handle it with your `~/error.vue`. (This also should decrease your total bundle size.)

👉 Compared to Nuxt 3.5.3, the minimal client bundle has decreased by ~0.7kB. Let's keep this up!

### 🔥 Fully static server components

It has been possible to use server components on static pages, but until now they would increase the payload size of your application. That is no longer true. We now store [rendered server components as separate files, which are preloaded before navigation](https://github.com/nuxt/nuxt/pull/21461).

👉 **This does rely on the new, richer JSON payload format**, so make sure you have not disabled this by setting `experimental.renderJsonPayloads` to false.

### 🎨 Better style inlining

If you're monitoring your metrics closely and have not turned off `experimental.inlineSSRStyles`, you should see more CSS inlined in your page, and a significantly external CSS file. We're now [better at deduplicating global CSS](https://github.com/nuxt/nuxt/pull/21573), particularly added by libraries like tailwind or unocss.

### 🎬 Animation controls

To give you more fine-grained control over your page/layout components, for example to create custom transitions with GSAP or other libraries, we now allow you to set [`pageRef` on `<NuxtPage>`](https://github.com/nuxt/nuxt/pull/19403) and [`layoutRef` on `<NuxtLayout`](https://github.com/nuxt/nuxt/pull/19465). These will get passed through to the underlying DOM elements.

### ✨ Automatic 'static' preset detection

Up to now, running `nuxt generate` produced the same output on every deployment provider, but with Nuxt 3.6 we now enable [static provider presets](https://github.com/nuxt/nuxt/pull/21655) automatically. That means if you are deploying a static build (produced with `nuxt generate`) to a supported provider (currently vercel and netlify with cloudflare and github pages coming soon) we'll prerender your pages with special support for that provider.

This means we can configure any route rules (redirects/headers/etc) that do not require a server function. So you should get the best of both worlds when deploying a site that doesn't require runtime SSR. It also unblocks use of [Nuxt Image](https://github.com/nuxt/image) on Vercel (with more potential for automatic provider integration coming soon).

### 💪 Increased type safety

We now have better support for server-specific `#imports` and augmentations if you are using the new `~/server/tsconfig.json` we shipped in Nuxt 3.5. So when importing from `#imports` in your server directory, you'll get IDE auto-completion for the right import locations in Nitro, and won't see Vue auto-imports like `useFetch` that are unavailable within your server routes.

You should now also have [type support for runtime Nitro hooks](https://github.com/nuxt/nuxt/pull/21666).

Finally, we have [removed more locations where objects had a default `any` type](https://github.com/nuxt/nuxt/pull/21700). This should improve type safety within Nuxt in a number of locations where unspecified types fell back to any:

* `RuntimeConfig`
* `PageMeta`
* `NuxtApp['payload']` (accessible now from `NuxtPayload` interface)
* `ModuleMeta`

You can find out more about how to [update your code](https://github.com/nuxt/nuxt/pull/21700) if this affects you in the original PR.

### ⚗️ Nitro 2.5 built-in

This release ships with new Nitro 2.5, which has a [whole list of exciting improvements](https://github.com/unjs/nitro/releases/tag/v2.5.0) that are worth checking out.

Of particular note is experimental support for streaming, which is also enabled by a [couple of changes](https://github.com/nuxt/nuxt/pull/21665) in Nuxt itself.

### 🛠️ New tools for module authors

This release brings a number of utilities for modules authors to easily [add type templates](https://github.com/nuxt/nuxt/pull/21331) and [assert compatibility](https://github.com/nuxt/nuxt/pull/21246) with a given version of _another_ module.

In addition, this release will finally unlock a new `nuxt/module-builder` mode that should improve type support for module authors. If you're a module author, you might consider following [these migration steps](https://github.com/nuxt/starter/pull/392) to try it out in the coming days.

## ✅ Upgrading

As usual, our recommendation for upgrading is to run:

```sh
nuxi upgrade --force
```

This will refresh your lockfile as well, and ensures that you pull in updates from other dependencies that Nuxt relies on, particularly in the unjs ecosystem.

## 👉 Changelog

[compare changes](https://github.com/nuxt/nuxt/compare/v3.5.3...v3.6.0)

<hr /><em>This discussion was created from the release <a href='https://github.com/nuxt/nuxt/releases/tag/v3.6.0'>v3.6.0</a>.</em>

**Comment by luc122c** [comment]
Hi There! I'm using [fontsource](https://fontsource.org/) to manage fonts in my project. To install a font, I install the package, then add it to the `css` key of `defineNuxtConfig` like so:  
```js
export default defineNuxtConfig({
  css: ["@fontsource/exo"].
  // ...
});
```
In the console, I now get the following warning: 
> **Warning**
> [plugin:ssr-styles] [nuxt] Cannot extract styles for @fontsource/exo. Its styles will not be inlined when server-rendering.          

I presume what it's telling me is that no styles can be extracted becasue it's essentially just a load of `@font-face` declarations. If this is the case, then is it safe to ignore this warning?

**Reply by danielroe** [reply]
You can safely ignore the warning. It will still be included in your CSS, just not inlined.

**Comment by Barbapapazes** [comment]
Can't wait to see what you've cooked for the CLI!

**Comment by localusercamp** [comment]
You are all doing a brilliant job. 
Thank you for that. 
🎉

Waiting for useFetch fabric =)

**Comment by zyhnbyyds** [comment]
Nice🎉🎉🎉

**Comment by huaiguoguo** [comment]
when will 3.7 be released?
In the coming week?
next month?

**Reply by danielroe** [reply]
You can see the timetable for upcoming minor releases here: https://github.com/nuxt/nuxt/milestones

There is normally also a patch release each week.

**Comment by so1ve** [comment]
What is nuxt-ng :eyes:

**Reply by Barbapapazes** [reply]
👉 https://www.npmjs.com/package/nuxi-ng

**Reply by so1ve** [reply]
@Barbapapazes Awesome! It's not open source (for now) right?

**Reply by Barbapapazes** [reply]
Yes, not ooen source but it sill ooen as was saying Daniel

**Comment by rsm23** [comment]
While updating I got errors because "typescript" module could not be found, added "typescript" as a dev dependency and the 
upgrade succeeded, but after the update the project could not be run because of this error

`[Vue warn]: <Suspense> slots expect a single root node.`

<img width="518" alt="Capture d’écran 2023-06-25 à 03 37 58" src="https://github.com/nuxt/nuxt/assets/2084708/da4dece3-c246-4120-a2d6-05f69b944318">

Anyone knows how to fix it?

**Reply by gitamgadtaula** [reply]
I am getting the same error, its frustrating that with almost every new releases of Nuxt these days, I always have to lose my time upgrading and downgrading it because it just doesn't work until a next minor patch is released.

**Reply by danielroe** [reply]
I'm sorry that this upgrade wasn't as smooth as it should have been for you.

Regarding the Suspense warning, would you open an issue so we can investigate and discuss?

**Reply by SpiritusDeos** [reply]
Same here when trying running, I get this error

```
Cannot start nuxt:  Cannot find module 'typescript'     

  Require stack:
  - node_modules/@typescript-eslint/typescript-estree/dist/convert.js
  - node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js
  - node_modules/@typescript-eslint/typescript-estree/dist/parser.js                                                                                                                                     11:46:46 AM
  - node_modules/@typescript-eslint/typescript-estree/dist/index.js
  at Module._resolveFilename (node:internal/modules/cjs/loader:1047:15)
  at Module._load (node:internal/modules/cjs/loader:893:27)
  at Module.require (node:internal/modules/cjs/loader:1113:19)
  at require (node:internal/modules/cjs/helpers:103:18)
  at Object.<anonymous> (node_modules/@typescript-eslint/typescript-estree/dist/convert.js:29:25)
  at Module._compile (node:internal/modules/cjs/loader:1226:14)
  at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
  at Module.load (node:internal/modules/cjs/loader:1089:32)
  at Module._load (node:internal/modules/cjs/loader:930:12)
  at Module.require (node:internal/modules/cjs/loader:1113:19)
  at require (node:internal/modules/cjs/helpers:103:18)
  at Object.<anonymous> (node_modules/@typescript-eslint/typescript-estree/dist/ast-converter.js:4:19)
  at Module._compile (node:internal/modules/cjs/loader:1226:14)
  at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
  at Module.load (node:internal/modules/cjs/loader:1089:32)
  at Module._load (node:internal/modules/cjs/loader:930:12)
  at Module.require (node:internal/modules/cjs/loader:1113:19)
  at require (node:internal/modules/cjs/helpers:103:18)
  at Object.<anonymous> (node_modules/@typescript-eslint/typescript-estree/dist/parser.js:8:25)
  at Module._compile (node:internal/modules/cjs/loader:1226:14)
  at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
  at Module.load (node:internal/modules/cjs/loader:1089:32)
  at Module._load (node:internal/modules/cjs/loader:930:12)
  at Module.require (node:internal/modules/cjs/loader:1113:19)
  at require (node:internal/modules/cjs/helpers:103:18)
  at Object.<anonymous> (node_modules/@typescript-eslint/typescript-estree/dist/index.js:18:16)
  at Module._compile (node:internal/modules/cjs/loader:1226:14)
  at Module._extensions..js (node:internal/modules/cjs/loader:1280:10)
  at Module.load (node:internal/modules/cjs/loader:1089:32)
  at Module._load (node:internal/modules/cjs/loader:930:12)
  at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:169:29)
  at ModuleJob.run (node:internal/modules/esm/module_job:194:25)
```

**Reply by danielroe** [reply]
The typescript issue should be resolved in the [edge channel](https://nuxt.com/docs/guide/going-further/edge-channel/) via https://github.com/nuxt/nuxt/pull/21729, or in the next release.

For now, you can add it as a dev dependency.

**Comment by nwittwer** [comment]
The possibilities around [Animation controls](https://nuxt.com/blog/v3-6#animation-controls) in v3.6 sound promising, but it's hard to understand how to use `pageRef` and `layoutRef` with a library like GSAP in practice. Are there any demos/examples? Would `middleware` be the right place to try to work with these for animation purposes? The docs added in the two PRs for `pageRef` and `layoutRef` are not very helpful in this regard.

**Reply by nwittwer** [reply]
Hi @danielroe, any resources or examples of how to use these^? Any pointers would be really helpful.
